ethcore/src/client/client.rs

```diff
@@ -255,14 +255,9 @@ impl Importer {
 		// could be retracted in import `k+1`. This is why to understand if after all inserts
 		// the block is enacted or retracted we iterate over all routes and at the end final state
 		// will be in the hashmap
-		let map = import_results.iter().fold(HashMap::new(), |mut map, route| {
-			for hash in &route.enacted {
-				map.insert(hash.clone(), true);
-			}
-			for hash in &route.retracted {
-				map.insert(hash.clone(), false);
-			}
-			map
+		let map = import_results.iter().flat_map(|route| {
```

Comment:
I think that won't work. The comment above explains what we tried to achieve with the previous code.
We need to remove duplicated blocks and make sure we only take last status of a block (i.e. it could go from `enacted` to `retracted` and then back to `enacted` in `ImportRoute`) - with current approach later on, after partition the block will end up in both `enacted` and `retracted`.

Reply:
@tomusdrw I think it probably doesn't matter a lot. `calculate_enacted_retracted` is only used in two places in `client.rs` (L334 and L2080). The only usage for calculated `enacted` and `retracted` is to pass it to `notify.new_blocks`.

I can't think of a good way to remove duplicate while preserving the actual route history -- it looks like the simplest way is to call `tree_route` for another time with the first and last block, but that may be already too expensive?

So I'm thinking about, it maybe a better idea to change the signature of `ChainNotify`? Instead of keep `enacted: &[H256], retracted: &[H256]`, we change it to `route: &[(H256, Status::Enacted/Status::Retracted)]`. Many listeners would just need the later, like the RPC `pubsub` mod, where it's expected to re-push headers in every re-org. If listeners need the former, it's also trivial to construct from the later. What do yo think?

Reply:
^^ Makes sense to me, that would be the best I suppose, but we need to review all cases where `enacted` and `retratected` is used to make sure we are not breaking anything by introducing that change.